### ISSUE TITLE
Too many open (leaking?) File Descriptors and "cannot import name SHA"

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4141]
|<img alt="cb's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| cb reported|
|-|-|
|Trac ID|trac#4141|
|Type|defect|
|Created|2009-11-30 23:56:15Z|

Hello
 my name is Christian Bahls
 i am an anti-censorship activist from germany
 
 we have this new law on internet-censorhip that
 would oblige providers to filter dns

 so i set out and implement a dns-(json+http)-dns-tunnel

My Resolver looks like this:
    resolvers=[]
    cacheResolver = cache.CacheResolver(verbose=True)
    resolvers.append(cacheResolver)
    hostResolver = hosts.Resolver(file='named.root.txt')
    #clientResolver = client.Resolver(servers=[('139.30.8.7',53)])
    rootResolver = root.bootstrap(hostResolver)
    resolvers.append(rootResolver)
    #resolvers.append(hostResolver)
    #resolvers.append(clientResolver)
    theResolver = resolve.ResolverChain(resolvers)

it gets called like this:
    def render_GET(self,request):
        DNSLogger.info("REQUEST: %s" % (request))
        d=theResolver.typeToMethod[self.typeKey](self.query)
        d.addCallback(self.returnResults,request)
        d.addErrback(self.returnError,request)
        #request.notifyFinish().addErrback(self.cancelRequest, d, request)
        return NOT_DONE_YET

   where typeKey would be '1' for type 'IN'

Under heavy Pressure .. I get:

Unhandled error in Deferred:
Traceback (most recent call last):
  File "/usr/lib/python2.5/site-packages/Twisted-9.0.0-py2.5-linux-x86_64.egg/twisted/internet/defer.py", line 238, in callback
    
  File "/usr/lib/python2.5/site-packages/Twisted-9.0.0-py2.5-linux-x86_64.egg/twisted/internet/defer.py", line 307, in _startRunCallbacks
    
  File "/usr/lib/python2.5/site-packages/Twisted-9.0.0-py2.5-linux-x86_64.egg/twisted/internet/defer.py", line 323, in _runCallbacks
    
  File "/usr/lib/python2.5/site-packages/Twisted-9.0.0-py2.5-linux-x86_64.egg/twisted/internet/defer.py", line 625, in gotResult
    
--- <exception caught here> ---
  File "/usr/lib/python2.5/site-packages/Twisted-9.0.0-py2.5-linux-x86_64.egg/twisted/internet/defer.py", line 600, in _deferGenerator
    
  File "/usr/lib/python2.5/site-packages/Twisted-9.0.0-py2.5-linux-x86_64.egg/twisted/names/root.py", line 120, in discoverAuthority
    
  File "/usr/lib/python2.5/site-packages/Twisted-9.0.0-py2.5-linux-x86_64.egg/twisted/names/root.py", line 67, in lookupNameservers
    
  File "/usr/lib/python2.5/site-packages/Twisted-9.0.0-py2.5-linux-x86_64.egg/twisted/names/root.py", line 31, in retry
    
  File "/usr/lib/python2.5/site-packages/Twisted-9.0.0-py2.5-linux-x86_64.egg/twisted/names/dns.py", line 1730, in query
    
  File "/usr/lib/python2.5/site-packages/Twisted-9.0.0-py2.5-linux-x86_64.egg/twisted/names/dns.py", line 1577, in pickID
    
  File "/usr/lib/python2.5/site-packages/Twisted-9.0.0-py2.5-linux-x86_64.egg/twisted/names/dns.py", line 44, in randomSource
    
  File "/usr/lib/python2.5/site-packages/Twisted-9.0.0-py2.5-linux-x86_64.egg/twisted/python/randbytes.py", line 113, in secureRandom
    
  File "/usr/lib/python2.5/site-packages/Twisted-9.0.0-py2.5-linux-x86_64.egg/twisted/python/randbytes.py", line 92, in _cryptoRandom
    
  File "/usr/lib/python2.5/site-packages/Crypto/Util/randpool.py", line 57, in __init__
    
exceptions.ImportError: cannot import name SHA

####### AND AFTERWARDS #########
Unhandled error in Deferred:
Traceback (most recent call last):
  File "/usr/lib/python2.5/site-packages/Twisted-9.0.0-py2.5-linux-x86_64.egg/twisted/internet/defer.py", line 600, in _deferGenerator
    
  File "/usr/lib/python2.5/site-packages/Twisted-9.0.0-py2.5-linux-x86_64.egg/twisted/names/root.py", line 120, in discoverAuthority
    
  File "/usr/lib/python2.5/site-packages/Twisted-9.0.0-py2.5-linux-x86_64.egg/twisted/names/root.py", line 67, in lookupNameservers
    
  File "/usr/lib/python2.5/site-packages/Twisted-9.0.0-py2.5-linux-x86_64.egg/twisted/names/root.py", line 31, in retry
    
--- <exception caught here> ---
  File "/usr/lib/python2.5/site-packages/Twisted-9.0.0-py2.5-linux-x86_64.egg/twisted/names/dns.py", line 1725, in query
    
  File "/usr/lib/python2.5/site-packages/Twisted-9.0.0-py2.5-linux-x86_64.egg/twisted/names/dns.py", line 1666, in startListening
    
  File "/usr/lib/python2.5/site-packages/Twisted-9.0.0-py2.5-linux-x86_64.egg/twisted/internet/posixbase.py", line 248, in listenUDP
    
  File "/usr/lib/python2.5/site-packages/Twisted-9.0.0-py2.5-linux-x86_64.egg/twisted/internet/udp.py", line 83, in startListening
    
  File "/usr/lib/python2.5/site-packages/Twisted-9.0.0-py2.5-linux-x86_64.egg/twisted/internet/udp.py", line 91, in _bindSocket
    
twisted.internet.error.CannotListenError: Couldn't listen on any:0: (24, 'Too many open files').

I saw a similar Bug arround here ..
.. but it was supposed to be fixed?

.. as i have only the Deferred to play around with
 i am not so sure how to get rid of the open UDP ports

yours Christian Bahls

Attachments:

* [snippets.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a9f/a9f132052b56a9395be4541057f919498e500a44/fa572849ef8a6623aff30f7f67ecc69a7cab3cdc.py) (1673 bytes) - added by cb on 2009-12-01 00:44:36Z - snippets from twisted/names/root.py
* [howdoicallit.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a9f/a9f132052b56a9395be4541057f919498e500a44/226df63606956b5f62b73b4e60b5d70b35afac04.py) (5200 bytes) - added by cb on 2009-12-01 01:03:30Z - How the service is run .. removed all the specific stuff ..
* [named.root.txt](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a9f/a9f132052b56a9395be4541057f919498e500a44/84895fb46efcb0aecd5a29395367fb945c444e11.txt) (663 bytes) - added by cb on 2009-12-01 13:38:34Z - named.root for bootstrapping the root resolver
* [showtheleakage.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/a9f/a9f132052b56a9395be4541057f919498e500a44/abf0b6069fd30d2a64f3da35c60138313aa0c319.py) (1226 bytes) - added by cb on 2009-12-01 14:18:46Z - This file calles rootResolver repeatedly .. and shows the leakage even faster

<details><summary>Searchable metadata</summary>

```
trac-id__4141 4141
type__defect defect
reporter__cb cb
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__duplicate duplicate
component__names names
keywords__ 
time__1259625375000000 1259625375000000
changetime__1260128012000000 1260128012000000
version__None None
owner__cb cb
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|<img alt="cb's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_1"></a>|cb commented|
|-|-|

Well .. i had a short look into twisted.names.root

there are two sources for DNSDatagramProtocol leaks i think

def lookupNameservers(host, atServer, p=None):
    # print 'Nameserver lookup for', host, 'at', atServer, 'with', p
    if p is None:
        p = dns.DNSDatagramProtocol(_DummyController())
        p.noisy = False
    return retry(
        (1, 3, 11, 45),                     # Timeouts
        p,                                  # Protocol instance
        (atServer, dns.PORT),               # Server to query
        [dns.Query(host, dns.NS, dns.IN)]   # Question to ask
    )

def lookupAddress(host, atServer, p=None):
    # print 'Address lookup for', host, 'at', atServer, 'with', p
    if p is None:
        p = dns.DNSDatagramProtocol(_DummyController())
        p.noisy = False
    return retry(
        (1, 3, 11, 45),                     # Timeouts
        p,                                  # Protocol instance
        (atServer, dns.PORT),               # Server to query
        [dns.Query(host, dns.A, dns.IN)]    # Question to ask
    )

lookupNamerservers() is being called from

def discoverAuthority(host, roots, cache=None, p=None):
    msg = defer.waitForDeferred(lookupNameservers(soFar, authority, p))

with p being 'None'

discoverAuthority() is being called on every _lookup()

    def _lookup(self, name, cls, type, timeout):
        d = discoverAuthority(name, self.hints
            ).addCallback(self.discoveredAuthority, name, cls, type, timeout
            )
        return d

additionally discoveredAuthority() initializes a Resolver for every request:

    def discoveredAuthority(self, auth, name, cls, type, timeout):
        from twisted.names import client
        q = dns.Query(name, type, cls)
        r = client.Resolver(servers=[(auth, dns.PORT)])
        d = r.queryUDP([q], timeout)
        d.addCallback(r.filterAnswers)
        return d

perhaps it could be done as in lookupAddress() 

still this does not solve the leaking of protocols
what to do about that?

yours
  Christian Bahls

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to cb|
|-|-|

I tried running `howdoicallit.py`, but I get this on stdout:

```
Unhandled error in Deferred:
Traceback (most recent call last):
Failure: twisted.names.error.DomainError: m.root-servers.net
Unhandled error in Deferred:
Traceback (most recent call last):
Failure: twisted.names.error.DomainError: l.root-servers.net
Unhandled error in Deferred:
Traceback (most recent call last):
Failure: twisted.names.error.DomainError: k.root-servers.net
Unhandled error in Deferred:
Traceback (most recent call last):
Failure: twisted.names.error.DomainError: j.root-servers.net
Unhandled error in Deferred:
Traceback (most recent call last):
Failure: twisted.names.error.DomainError: i.root-servers.net
Unhandled error in Deferred:
Traceback (most recent call last):
Failure: twisted.names.error.DomainError: h.root-servers.net
Unhandled error in Deferred:
Traceback (most recent call last):
Failure: twisted.names.error.DomainError: g.root-servers.net
Unhandled error in Deferred:
Traceback (most recent call last):
Failure: twisted.names.error.DomainError: f.root-servers.net
Unhandled error in Deferred:
Traceback (most recent call last):
Failure: twisted.names.error.DomainError: e.root-servers.net
Unhandled error in Deferred:
Traceback (most recent call last):
Failure: twisted.names.error.DomainError: d.root-servers.net
Unhandled error in Deferred:
Traceback (most recent call last):
Failure: twisted.names.error.DomainError: c.root-servers.net
Unhandled error in Deferred:
Traceback (most recent call last):
Failure: twisted.names.error.DomainError: b.root-servers.net
Unhandled error in Deferred:
Traceback (most recent call last):
Failure: twisted.names.error.DomainError: a.root-servers.net
```

And I don't see how I'm supposed to interact with the web server.  Can you provide a self-contained example, or instructions on how to use this example?

**Comment by twisted-trac** [comment]
|<img alt="cb's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|cb commented|
|-|-|

of course .. you will need named.root.txt for bootstrapping the rootResolver
with the hostResolver


you interact by getting
*http://127.0.0.1:8881/IN/A/whatever 
*http://127.0.0.1:8881/IN/PTR/4.3.2.1.in-addr.arpa
*http://127.0.0.1:8881/IN/ALL_RECORDS/whatever

but the Problem isn't with the Webserver ..

.. the Problem is that the rootResolver randomizes
it's port by generating dns.DNSDatagramProtocol()s all the Time
and never goes arround to actually get rid of them

so you can get the same Effect by just calling a
rootResolvers _lookup() routine repeatedly

i will attach an example ..
.. it will need the named.root.txt as a hint for bootstrapping the rootResolver

(ah .. and catching KeyboardInterrupt does not work as expected .. sorry for the bad failure handling)

Christia

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set status to `closed`|
|-|-|

Ah, okay.  This is a duplicate of [#10648](https://github.com/twisted/twisted/issues/10648).

**Comment by twisted-trac** [comment]
|<img alt="cb's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|cb set status to `reopened`|
|-|-|

I am not using Windows ..
.. this is not the same Bug

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun set status to `closed`|
|-|-|

The mention of Windows on [#10648](https://github.com/twisted/twisted/issues/10648) is a mistake, it applies to all platforms.

**Comment by twisted-trac** [comment]
|<img alt="cb's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|cb set status to `reopened`|
|-|-|

No .. the solution in [#10648](https://github.com/twisted/twisted/issues/10648) does not help me .. as it uses the ThreadedResolver() which in turn calls socket.gethostbyname

gethostbyname will only end up asking the root.Resolver i am just trying to instantiate

the root.Resolver should be more careful with the resources being used

see: http://twistedmatrix.com/trac/ticket/4141#comment:4

if this is an error that will never be resolved than just tell me right away .. than i will stop bothering you with my requests ..

Christian

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

[#10648](https://github.com/twisted/twisted/issues/10648) is still open, it has no solution yet.  The `ThreadedResolver` change is just a comment someone put on the ticket.  It wasn't an accepted solution, so the ticket is still open.

You're right that `root.Resolver` isn't cleaning up its resources.  Fixing that problem is just what I think the solution to [#10648](https://github.com/twisted/twisted/issues/10648) should be.

**Comment by twisted-trac** [comment]
|<img alt="cb's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|cb commented|
|-|-|

to me it looks as if the asynchronous Eventhandling has a design-bug

it boils down to this:
 * how and when to throw away DNSDatagramProtocol-Instances
 (which would mean to somehow keep track of their usage)

yours
  Christian

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_11"></a>|@thijstriemstra commented|
|-|-|

[#10648](https://github.com/twisted/twisted/issues/10648) mentions this ticket as duplicate?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun set status to `closed`|
|-|-|

Yes.  The fix for [#10648](https://github.com/twisted/twisted/issues/10648) will take care of this too.
